tower-ready-cache/src/lib.rs

```diff
@@ -0,0 +1,276 @@
+                    let _cancel = self.cancelations.remove(&key);
+                    debug_assert!(_cancel.is_some(), "missing cancelation");
+                    self.ready_services.insert(key, svc);
+                }
+                Err((key, UnreadyError::Canceled)) => {
+                    debug_assert!(!self.cancelations.contains_key(&key));
+                }
+                Err((key, UnreadyError::Inner(e))) => {
+                    debug!("dropping failed endpoint: {:?}", e);
+                    let _cancel = self.cancelations.swap_remove(&key);
+                    debug_assert!(_cancel.is_some());
+                }
+            }
+        }
+    }
+
+    /// Polls a service in the ready set to determine if it is still ready.
+    ///
+    /// If the service is no longer ready and there are no equivalent unready
+    /// services, it is moved back into the ready set and `Async::NotReady` is
+    /// returned.
+    ///
+    /// If the service errors, it is removed and dropped and the error is returned.
+    ///
+    /// Otherwise, `Async::Ready` is returned.
+    ///
+    /// Panics if the provided `index` is out of range.
+    pub fn poll_ready_index(&mut self, index: usize) -> Poll<(), Error> {
+        let (_, svc) = self
+            .ready_services
+            .get_index_mut(index)
+            .expect("index out of range");
+
+        match svc.poll_ready() {
+            Ok(Async::Ready(())) => Ok(Async::Ready(())),
+            Ok(Async::NotReady) => {
+                // became unready; so move it back there.
+                let (key, svc) = self
+                    .ready_services
+                    .swap_remove_index(index)
+                    .expect("invalid ready index");
+                // If a new version of this service has been added to the
+                // unready set, don't overwrite it.
+                if !self.unready_contains(&key) {
+                    self.push_unready(key, svc);
+                }
+                Ok(Async::NotReady)
+            }
+            Err(e) => {
+                // failed, so drop it.
+                let e = e.into();
+                debug!("evicting failed endpoint: {:?}", e);
+                self.ready_services
+                    .swap_remove_index(index)
+                    .expect("invalid ready index");
+                Err(e)
+            }
+        }
+    }
+
+    /// Swap-removes the indexed service from the ready set.
+    ///
+    /// Panics if the provided `index` is out of range.
+    pub fn process_ready_index<T, P>(&mut self, index: usize, process: P) -> T
+    where
+        P: FnOnce(&mut S) -> T,
+    {
+        let (key, mut svc) = self
+            .ready_services
+            .swap_remove_index(index)
+            .expect("index out of range");
+
+        let t = (process)(&mut svc);
+
+        // If a new version of this service has been added to the
+        // unready set, don't overwrite it.
```

Comment:
This case is weird to me. How would that happen?

Reply:
`ReadyCache::push_unready` may be called for a key that is already in the ready set and the ready instance will not be replaced until the unready instance becomes available.

Imagine that you get an update from service discovery that reconfigures a connection. I wouldn't want to be forced to tear down an old connection until the new connection is ready.